src/test/java/com/google/devtools/build/lib/analysis/ToolchainResolverTest.java

```diff
@@ -433,6 +437,102 @@ public void resolve_noMatchingPlatform() throws Exception {
         .isInstanceOf(NoMatchingPlatformException.class);
   }
 
+  @Test
+  public void unloadedToolchainContext_load() throws Exception {
+    addToolchain(
+        "extra",
+        "extra_toolchain_linux",
+        ImmutableList.of("//constraints:linux"),
+        ImmutableList.of("//constraints:linux"),
+        "baz");
+    rewriteWorkspace(
+        "register_toolchains('//extra:extra_toolchain_linux')",
+        "register_execution_platforms('//platforms:linux')");
+
+    useConfiguration("--platforms=//platforms:linux");
+    ResolveToolchainsKey key =
+        ResolveToolchainsKey.create("test", ImmutableSet.of(testToolchainType), targetConfigKey);
+
+    // Create the UnloadedToolchainContext.
+    EvaluationResult<ResolveToolchainsValue> result = createToolchainContextBuilder(key);
+    assertThatEvaluationResult(result).hasNoError();
+    UnloadedToolchainContext unloadedToolchainContext = result.get(key).unloadedToolchainContext();
+    assertThat(unloadedToolchainContext).isNotNull();
+
+    // Create the prerequisites.
+    ConfiguredTargetAndData toolchain = getConfiguredTargetAndData(
+        Label.parseAbsoluteUnchecked("//extra:extra_toolchain_linux_impl"), targetConfig);
+    OrderedSetMultimap<Attribute, ConfiguredTargetAndData> prerequisiteMap = OrderedSetMultimap.create();
+    prerequisiteMap.put(Attribute.attr(PlatformSemantics.RESOLVED_TOOLCHAINS_ATTR, LABEL_LIST).build(), toolchain);
+
+    ToolchainContext toolchainContext = unloadedToolchainContext.load(prerequisiteMap);
+    assertThat(toolchainContext).isNotNull();
+    assertThat(toolchainContext.forToolchainType(testToolchainType)).isNotNull();
+    assertThat(toolchainContext.forToolchainType(testToolchainType).hasField("data")).isTrue();
+    assertThat(toolchainContext.forToolchainType(testToolchainType).getValue("data")).isEqualTo("baz");
+  }
+
+  @Test
+  public void unloadedToolchainContext_load_withTemplateVariables() throws Exception {
+    // Add new toolchain rule that provides template variables.
+    Label variableToolchainType =
+        Label.parseAbsoluteUnchecked("//variable:variable_toolchain_type");
+    scratch.file(
+        "variable/variable_toolchain_def.bzl",
+        "def _impl(ctx):",
+        "  value = ctx.attr.value",
+        "  toolchain = platform_common.ToolchainInfo()",
+        "  template_variables = platform_common.TemplateVariableInfo({'VALUE': value})",
```

Comment:
So the [docs](https://docs.bazel.build/versions/master/skylark/lib/platform_common.html#TemplateVariableInfo) describe this as:

> The provider used to retrieve the provider that contains the template variables defined by a particular toolchain

Can you parse this statement for me? We have to go through two providers to get the final info?

Reply:
That is pretty ugly. Good thing I'm auditing the platform-related documentation this week.

TemplateVariableInfo contains a map of variable names to values, and can be used for automatic make variable expansion (ie, replacing ${FOO} in genrule's cmd, or various other places we handle make var expansion).